### PR TITLE
[Example] Temporal Graph Neural Network

### PR DESCRIPTION
**Pull request by Ericcsr**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/tgn/data.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    # copy back edge and node data
+    for ek in edge_keys:
+        buf_graph.edata[ek] = g.edata[ek][sorted_idx]
+
+    # Since node index unchanged direct copy
+    for nk in node_keys:
+        buf_graph.ndata[nk] = g.ndata[nk]
+    return buf_graph
+
+# Probe ahead to get the valid spliting point
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data.py:179]: What is a splitting point?

File: examples/pytorch/tgn/data.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+def probe_division(g,div):
+    for i in range(div,-1,-1):
+        ret = (g.edata[dgl.EID]==i).int().argmax()
+        if ret:
+            return ret
+
+# TODO: Define the dataloader
+class DictNode:
+    def __init__(self,parent,NIDdict=None):
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data.py:188]: Need comments on parent and NIDdict.

File: examples/pytorch/tgn/data.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        #num_nodes = g.num_nodes()
+        train_div = int(0.7*num_edges)
+        valid_div = int(0.85*num_edges)
+
+        self.nn_test_g  = dgl.edge_subgraph(self.g,range(valid_div,num_edges))
+        self.nn_test_g = temporal_sort(self.nn_test_g,'timestamp')
+
+        self.d_nn_test_g= DictNode(parent=self.d_g,NIDdict=self.nn_test_g.ndata[dgl.NID])
+        # New Node random non-repeated choice
+        np.random.seed(2020)
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data.py:224]: Usually random seed is set at main.py file.

File: examples/pytorch/tgn/data.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        test_nodes = self.nn_test_g.ndata[dgl.NID].numpy()
+        #print(test_nodes.shape)
+        n_test_nodes = len(test_nodes)
+        # We Need to get node subgraph from previous training graph
+        #print(int(0.1*n_test_nodes))
+        new_node_idxs = np.random.choice(test_nodes,replace=False,size=int(0.1*n_test_nodes))
+        non_new_node_idxs = np.delete(self.g.nodes().numpy(),new_node_idxs)
+        #print(new_node_idxs)
+        # Need to get the training and validation subgraph
+        self.nn_mask_g   = dgl.node_subgraph(self.g,non_new_node_idxs)
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data.py:235]: What is the 'nn' here means?

File: examples/pytorch/tgn/data.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        Here batch is the minimum unit for memory update
+        The gradient update should use hyperbatch
+        Since graph here is already subgraph. No need offset
+        Maybe need negative sampling mechanism? Only on edge index domain
+        '''
+        graph = self.graph_dict[mode]
+        p_dict= self.dict_dict[mode]
+        done = False
+        #print(self.g.edges()[1].shape)
+        src_list = graph.edges()[0][self.batch_cnt*self.batch_size:min((self.batch_cnt+1)*self.batch_size,graph.num_edges())]
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data.py:280]: Line is too long. Need refactoring.

File: examples/pytorch/tgn/modules.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+# TODO: Implement Time encoder
+class TimeEncode(torch.nn.Module):
+  # Time Encoding proposed by TGAT
+  def __init__(self, dimension):
+    super(TimeEncode, self).__init__()
+
+    self.dimension = dimension
+    self.w = torch.nn.Linear(1, dimension)
+
+    self.w.weight = torch.nn.Parameter((torch.from_numpy(1 / 10 ** np.linspace(0, 9, dimension)))
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:52]: Why this formula?

File: examples/pytorch/tgn/modules.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        return self.last_update_t[node_idxs]
+
+    def detach_memory(self):
+        # Since the memory update itself will not receive gradient but will be propagated to the next batched iteration
+        self.memory.detach_()
+        # How to detach edge weights since they are in graph
+
+# TODO: Check about the node id matching problem
+class MemoryOperation(nn.Module):
+    def __init__(self,updater_type, memory, e_feat_dim, temporal_dim):
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:102]: Comments on each parameter.

File: examples/pytorch/tgn/modules.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        super(MemoryOperation,self).__init__()
+        updater_dict = {'gru':nn.GRUCell,'rnn':nn.RNNCell}
+        self.memory = memory
+        memory_dim = self.memory.hidden_dim
+        self.message_dim = memory_dim+memory_dim+e_feat_dim+temporal_dim
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:107]: Why message_dim is computed like this?

File: examples/pytorch/tgn/modules.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        self.updater = updater_dict[updater_type](input_size=self.message_dim,
+                                                  hidden_size=memory_dim)
+        self.memory = memory
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:110]: Duplicate with line 105.

File: examples/pytorch/tgn/modules.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        self.temporal_encoder = TimeEncode(temporal_dim)
+
+    # Here assume g is a subgraph from each iteration
+    def stick_feat_to_graph(self,g):
+        # How can I ensure order of the node ID
+        g.ndata['timestamp'] = self.memory.last_update_t[g.ndata[dgl.NID]]
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:116]: This mapping is confusing.

File: examples/pytorch/tgn/modules.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        # Edge softmax
+        graph.edata['sa'] = edge_softmax(graph,graph.edata['a'])/(self._out_feats**0.5)
+
+        # Update dst node Here msg_fn include edge feature
+        graph.update_all(self.msg_fn,fn.sum('attn','agg_u'))
+
+        rst = graph.dstdata['agg_u']
+        # Implement skip connection
+        rst = self.merge(rst.view(-1,self._num_heads*self._out_feats),graph.dstdata['s'])
+        return rst
```
Comment (sneakerkg) [inline examples/pytorch/tgn/modules.py:226]: Usually we need an empty line at the end of each file.

File: logging.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+Hello World!!
```
Comment (sneakerkg) [inline logging.txt:1]: Remove this file.

File: examples/pytorch/tgn/train_gru.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
+import dgl
```
Comment (sneakerkg) [inline examples/pytorch/tgn/train_gru.py:1]: GRU and memory are trained separately?

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/tgn/data_preprocess.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        gs,_ = dgl.load_graphs('./data/{}.bin'.format(dataset))
+        g = gs[0]
+    return g
+
+def TemporalWikipediaDataset():
+    # Download the dataset
+    return TemporalDataset('wikipedia')
+
+def TemporalRedditDataset():
+    return TemporalDataset('reddit')
```
Comment (sneakerkg) [inline examples/pytorch/tgn/data_preprocess.py:130]: In some python format guide, we can add an empty line at the end of each file.

File: examples/pytorch/tgn/logging.txt [hunk 15]
```diff
@@ -0,0 +1,4 @@
+Epoch: 0; Training Loss: 38642.09296107292 | Validation AP: 0.988 AUC: 0.991
```
Comment (sneakerkg) [inline examples/pytorch/tgn/logging.txt:1]: Delete logging file.

File: examples/pytorch/tgn/train.py [hunk 16]
```diff
@@ -0,0 +1,243 @@
+import dgl
+import torch
+import numpy as np
+from tgn import TGN
+from data_preprocess import TemporalWikipediaDataset,TemporalRedditDataset,TemporalDataset
+from dataloading import FastTemporalEdgeCollator,NegativeSampler,FastTemporalSampler,TemporalEdgeDataLoader,temporal_sort,TemporalSampler,TemporalEdgeCollator
```
Comment (sneakerkg) [inline examples/pytorch/tgn/train.py:6]: If you import a lot of things from a file, a suggested way is:
from utils import (AA, BB, CC,
                             DD, EE, FF)

File: logging.txt (hunk 12 above)
Comment (sneakerkg) [reply logging.txt:1]: We still have this file?

File: examples/pytorch/tgn/data_preprocess.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+            feat_l.append(feat)
+    return pd.DataFrame({'u': u_list,
+                         'i': i_list,
+                         'ts': ts_list,
+                         'label': label_list,
+                         'idx': idx_list}), np.array(feat_l)
+
+
+# Re index nodes for DGL convience
+def reindex(df, bipartite=True):
```
Comment (WangXuhongCN) [inline examples/pytorch/tgn/data_preprocess.py:50]: perhaps we need a non-bipartite reindex module, like my [APAN code ](https://github.com/WangXuhongCN/APAN/blob/0474a5a5bc2729f0550d8454be6772f1ff92e8b7/preprocess/preprocess_csv.py#L104)

File: examples/pytorch/tgn/dataloading.py [hunk 18]
```diff
@@ -0,0 +1,731 @@
...
+        result = super().collate(items)
+        # Copy the feature from parent graph
+        _pop_subgraph_storage(result[1], self.g)
+        _pop_subgraph_storage(result[2], self.g)
+        _pop_blocks_storage(result[-1], self.g_sampling)
+        return result
+
+
+class TemporalEdgeDataLoader(dgl.dataloading.EdgeDataLoader):
+    """ TemporalEdgeDataLoader is a iteratable object to generate blocks for temporal embedding
```
Comment (sneakerkg) [inline examples/pytorch/tgn/dataloading.py:238]: a -> an

File: examples/pytorch/tgn/dataloading.py [hunk 19]
```diff
@@ -0,0 +1,731 @@
...
+
+# ====== Fast Mode ======
+
+# Part of code comes from PyG library
+# https://github.com/rusty1s/pytorch_geometric
+
+
+class FastTemporalSampler(BlockSampler):
+    """Temporal Sampler which implemented with a fast query lookup table. Sample
+    Temporal and computationally depending subgraph.
```
Comment (sneakerkg) [inline examples/pytorch/tgn/dataloading.py:297]: Temporal -> temporal

File: examples/pytorch/tgn/dataloading.py [hunk 20]
```diff
@@ -0,0 +1,731 @@
...
+
+    The sampler maintains a lookup table of most current k neighbors of each node
+    each time, the sampler need to be updated with new edges from incoming batch to
+    update the lookup table.
+
+    Parameters
+    ----------
+    g : dgl.Heterograph
+        graph to be sampled here it which only exist to provide feature and data reference
```
Comment (sneakerkg) [inline examples/pytorch/tgn/dataloading.py:306]: grammer

File: examples/pytorch/tgn/dataloading.py [hunk 21]
```diff
@@ -0,0 +1,731 @@
...
+
+        # Precompute the CSR and CSC representations so each subprocess does not
+        # duplicate.
+        if dataloader_kwargs.get('num_workers', 0) > 0:
+            g.create_formats_()
+
+# ====== Fast Mode ======
+
+# Part of code comes from PyG library
+# https://github.com/rusty1s/pytorch_geometric
```
Comment (sneakerkg) [inline examples/pytorch/tgn/dataloading.py:292]: Please given specific file name here. We just don't want to point to the whole PyG repo :)

File: examples/pytorch/tgn/dataloading.py [hunk 22]
```diff
@@ -0,0 +1,731 @@
...
+        self.ts = 0
+        self.frontiers = [None for _ in range(len(fanouts))]
+
+    def sample_frontier(self, block_id, g, seed_nodes):
+        '''
+        Deleting the the edges that happen after the current timestamp, then use a simple topk edge sampling by timestamp.
+        '''
+        fanout = self.fanouts[block_id]
+        # List of neighbors to sample per edge type for each GNN layer, starting from the first layer.
+        g = dgl.in_subgraph(g, seed_nodes)  # 只包含seed_node的g（不考虑邻居吗？）
```
Comment (sneakerkg) [inline examples/pytorch/tgn/dataloading.py:625]: Use English comments.

File: examples/pytorch/tgn/logging.txt [hunk 23]
```diff
@@ -0,0 +1,13 @@
+Epoch: 0; Training Loss: 43824.699464440346 | Validation AP: 0.977 AUC: 0.984
```
Comment (sneakerkg) [inline examples/pytorch/tgn/logging.txt:1]: Remove this file.

File: examples/pytorch/tgn/tgn.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
+import torch.nn as nn
+import dgl
+
+from modules import MemoryModule, MemoryOperation, TemporalGATConv, MsgLinkPredictor
+import copy
```
Comment (sneakerkg) [inline examples/pytorch/tgn/tgn.py:5]: Put this import to the first line.

File: logging.txt [hunk 25]
```diff
@@ -0,0 +1,5 @@
...
+Hello World!!
```
Comment (sneakerkg) [inline logging.txt:2]: Seems this file still exists.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/tgn/dataloading.py [hunk 26]
```diff
@@ -0,0 +1,731 @@
...
+        If 'uniform' then uniform randomly sample k nodes
+
+    k : int
+        maximum number of neighors to sampler
+
+        default 10 neighbors as paper stated
+
+    Examples
+    ----------
+    Please refers to examples/pytorch/tgn/train.py
```
[the remark this answers is not in the document]
Comment (Ericcsr) [reply examples/pytorch/tgn/dataloading.py:43]: Simple Mode can, others cannot

**Review by sneakerkg — APPROVED** [review]
